Despair-Games/poketernity
# [P3][Sprite] Fix game not waiting for variant data to finish loading

## What are the changes the user will see?

The game will no longer show the non shiny version of a Pokemon while variant data is still loading
- in starter select
- in ME intros


## Why am I making these changes?

Issue I introduced in PR 4827 (Fix ME Intro Visuals for shinies), it looks bad




## What are the changes from a developer perspective?

<sub> don't have it in me to convert all of asset loading code to using async/await, that's definitely for another PR

- `loadPokemonVariantAssets` now returns a promise
- `PokemonSpecies.loadAssets` waits for the variant loading to be done before resolving, like it did before PR 4827
- ME Intro visuals uses Promise.All to wait on all variant loading requests



## Screenshots/Videos

I added a several seconds delay to the loading so it's easier to see. But the issue is quite noticeable in Starter Select without that depending on device/network performance

<details><summary>Starter Select:</summary>

Before, the non shiny form shows before the variant finishes loading

https://github.com/user-attachments/assets/a95bfa31-1dcc-43f9-a2f7-160ba7537951

After, the sprite only shows once everything is loaded

https://github.com/user-attachments/assets/13a4b495-c7d8-4da6-8659-12ca79afc27d

</details>


<details><summary>ME Intro:</summary>

Before, the non shiny form shows before the variant finishes loading

https://github.com/user-attachments/assets/b71f90ee-513e-47b2-95ae-32d78ce47db2

After, the transition to ME doesn't happen until everything is loaded

https://github.com/user-attachments/assets/42c1cf06-c26c-4b3d-82b2-2da56e8a2180

</details>




## How to test the changes?

It's a bit annoying to test, just load starter select and play with shiny overrides and make sure there are no issues I guess

<details><summary>To add the delay in asset loading change 445 and 446 of BattleScene to:</summary>

```ts
            this.time.delayedCall(3000, () => {
              variantColorCache[spriteKey] = c;
              resolve();
            });
```
</details>

Relevant MEs: Dancing Lesson, Berries Abound, Fight or Flight, Uncommon Breed, Pokemon Salesman
<details><summary>Shiny overrides (they don't work for all these MEs)</summary>

```
  OPP_SHINY_OVERRIDE: true,
  OPP_VARIANT_OVERRIDE: X,
  MYSTERY_ENCOUNTER_RATE_OVERRIDE: 200,
  STARTING_WAVE_OVERRIDE: 11
```
</details>

<details><summary>To reproduce my testing scenario you can update the Pokemon Salesman ME:</summary>

Change line 94 to 
`pokemon = new PlayerPokemon(species, 5, 2, species.formIndex, undefined, true);`
or if you want a specific variant X: (0, 1 or 2)
`pokemon = new PlayerPokemon(species, 5, 2, species.formIndex, undefined, true, X);`

overrides:
```
  MYSTERY_ENCOUNTER_OVERRIDE: MysteryEncounterType.THE_POKEMON_SALESMAN,
  MYSTERY_ENCOUNTER_RATE_OVERRIDE: 255,
  STARTING_WAVE_OVERRIDE: 11
```

</details>



## Checklist

~~  - [ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?~~
~~  - [ ] Have I made sure that any UI change works for both UI themes (default and legacy)?~~

~~Are there any localization additions or changes? If so:~~
